packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,341 @@
+  name?: string;
+  /** Optional client version, defaults to '1.0.0' */
+  version?: string;
+}
+
+export async function createMCPClient(
+  config: MCPClientConfig,
+): Promise<MCPClient> {
+  const client = new MCPClient(config);
+  await client.init();
+  return client;
+}
+
+/**
+ * A lightweight MCP Client implementation
+ *
+ * The primary purpose of this client is tool conversion between MCP<>AI SDK
+ * but can later be extended to support other MCP features
+ *
+ * Tool parameters are automatically inferred from the server's JSON schema
+ * if not explicitly provided in the tools configuration
+ *
+ * Not supported:
+ * - Client options (e.g. sampling, roots) as they are not needed for tool conversion
+ * - Accepting notifications
+ */
+export class MCPClient {
+  private transport: MCPTransport;
+  private clientInfo: ClientConfiguration;
+  private requestMessageId = 0;
+  private responseHandlers: Map<
+    number,
+    (response: JSONRPCResponse | Error) => void
+  > = new Map();
+
+  constructor({
+    transport: transportConfig,
+    name = 'ai-sdk-mcp-client',
+    version = '1.0.0',
+  }: MCPClientConfig) {
+    this.transport = createMcpTransport(transportConfig);
+    this.transport.onclose = () => {
+      this.onclose();
+    };
+    this.transport.onerror = (error: Error) => {
+      this.onerror(error);
+    };
+    this.transport.onmessage = message => {
+      if (!('method' in message)) {
+        this.onresponse(message);
+      } else {
+        // This lightweight client implementation does not support
+        // receiving notifications or requests from server:
+        throw new MCPClientError({
+          message: 'Unsupported message type',
+        });
+      }
+    };
+    this.clientInfo = {
+      name,
+      version,
+    };
+  }
+
+  async init(): Promise<this> {
+    try {
+      await this.transport.start();
+
+      const result = await this.request({
+        request: {
+          method: 'initialize',
+          params: {
+            protocolVersion: LATEST_PROTOCOL_VERSION,
+            capabilities: {},
+            clientInfo: this.clientInfo,
+          },
+        },
+        resultSchema: InitializeResultSchema,
+      });
+
+      if (result === undefined) {
+        throw new MCPClientError({
+          message: 'Server sent invalid initialize result',
+        });
+      }
+
+      if (!SUPPORTED_PROTOCOL_VERSIONS.includes(result.protocolVersion)) {
+        throw new MCPClientError({
+          message: `Server's protocol version is not supported: ${result.protocolVersion}`,
+        });
+      }
+
+      if (!result.capabilities?.tools) {
+        throw new MCPClientError({
+          message: `Server does not support tools`,
+        });
+      }
+
+      // Complete initialization handshake:
+      await this.notification({
+        method: 'notifications/initialized',
+      });
+
+      return this;
+    } catch (error) {
+      void this.close();
```

Comment:
`void` is not needed? also, what are the implications of calling close here?

Reply:
noted on `void`! calling close here ensures we clean up the transport/connection and clear our response handlers